Pinta.Core/Classes/AsyncEffectRenderer.cs

```diff
@@ -181,78 +180,81 @@ void StartRender ()
 
 		// Start slave render threads.
 		int threadCount = settings.ThreadCount;
-		var slaves = new Thread[threadCount - 1];
-		for (int threadId = 1; threadId < threadCount; threadId++)
-			slaves[threadId - 1] = StartSlaveThread (renderId, threadId);
+		ThreadStart commonThreadStart = () => Render (sourceSurface, destSurface, renderInfos, renderId);
+		var slaves = Enumerable.Range (0, threadCount - 1).Select (_ => StartThread (commonThreadStart));
 
-		// Start the master render thread.
-		var master = new Thread (() => {
+		ThreadStart masterStart = commonThreadStart + (() => {
 
-			// Do part of the rendering on the master thread.
-			Render (renderId, 0);
+			// Part of the rendering is being done on the master thread, and we should add code to coordinate the slave threads
 
 			// Wait for slave threads to complete.
 			foreach (var slave in slaves)
 				slave.Join ();
 
 			// Change back to the UI thread to notify of completion.
 			GLib.Functions.TimeoutAdd (0, 0, () => {
-				HandleRenderCompletion ();
+				HandleRenderCompletion (sourceSurface, destSurface, renderInfos);
 				return false; // don't call the timer again
 			});
-		}) {
-			Priority = settings.ThreadPriority
-		};
-		master.Start ();
+
+			// Wait for slave threads to complete.
+			foreach (var slave in slaves)
+				slave.Join ();
+
+			// Change back to the UI thread to notify of completion.
+			GLib.Functions.TimeoutAdd (0, 0, () => {
+				HandleRenderCompletion (sourceSurface, destSurface, renderInfos);
+				return false; // don't call the timer again
+			});
+		});
+		_ = StartThread (masterStart);
 
 		// Start timer used to periodically fire update events on the UI thread.
 		timer_tick_id = GLib.Functions.TimeoutAdd (0, (uint) settings.UpdateMillis, () => HandleTimerTick ());
 	}
 
-	Thread StartSlaveThread (int renderId, int threadId)
+	Thread StartThread (ThreadStart threadStart)
 	{
-		var slave = new Thread (() => {
-			Render (renderId, threadId);
-		}) {
-			Priority = settings.ThreadPriority
-		};
-		slave.Start ();
-
-		return slave;
+		var newThread = new Thread (threadStart) { Priority = settings.ThreadPriority };
+		newThread.Start ();
+		return newThread;
 	}
 
+	internal readonly record struct TileRenderInfo (int TileIndex, RectangleI TileBounds);
+
 	// Runs on a background thread.
-	void Render (int renderId, int threadId)
+	void Render (
+		Cairo.ImageSurface sourceSurface,
+		Cairo.ImageSurface destSurface,
+		IEnumerable<TileRenderInfo> renderInfos,
+		int renderId)
 	{
-		// Fetch the next tile index and render it.
-		for (; ; ) {
-			int tileIndex = Interlocked.Increment (ref current_tile);
-			if (tileIndex >= total_tiles || cancel_render_flag)
-				return;
-			RenderTile (renderId, threadId, tileIndex);
+		foreach (var info in renderInfos) {
```

Comment:
I'm not 100% sure if I'm following the changes, but it seems like all of the threads are now going to each render all of the same tiles on top of each other?

In the previous code I think the `Interlocked.Increment (ref current_tile)` stuff was causing each thread to grab the next available tile that hadn't been rendered yet.
It might be better to see if using Parallel.For is possible to use here, but that's a much larger refactoring and this is quite large as-is

Reply:
Sorry for the mess. The original idea was to do a step-by-step refactoring, but the thing is that the state changes are so garbled that refactoring is not that practical. In fact I am thinking that I should just re-write it.

So don't worry, we could just discard these changes. On the bright side, it helped me understand what the `AsyncEffectRenderer` and the `LivePreviewManager` are doing.

I'm not 100% sure that I refactored it right, but if I did, then the threads were already rendering the same tiles on top of each other.

Reply:
`current_tile` was shared between the multiple threads, so I think each tile would only end up being rendered once by whichever thread happened to increment `current_tile` to that specific tile index